cdap-app-fabric/src/main/java/co/cask/cdap/common/twill/TwillAppLifecycleEventHandler.java

```diff
@@ -191,15 +188,13 @@ public void containerStopped(String runnableName, int instanceId, String contain
 
     String errorMessage = String.format("Container %s, instance %s stopped with exit status %d",
                                         containerId, instanceId, exitStatus);
-    LOG.error(errorMessage);
-
     switch(programRunId.getType()) {
       case WORKFLOW:
       case SPARK:
       case MAPREDUCE:
-        // For workflow, mapreduce, and spark, if there is an error, record the error state
-        if (stoppedContainer.compareAndSet(false, true)) {
-          programStateWriter.error(programRunId, new RuntimeException(errorMessage));
+        // For workflow, MapReduce, and spark, if there is an error, record the error state
+        if (completed.compareAndSet(false, true)) {
+          programStateWriter.error(programRunId, new Exception(errorMessage));
```

Comment:
For other program types, as per the current design, they'll never stop, even there is exception raised in container, as the AM would restart it (and eventually aborted the app if it keeps failing such that the time out abort kicks in). See in comment below in the `default:` block.